File: src/main/java/me/meiallu/luck/run/Lexer.java
package me.meiallu.luck.run;

import me.meiallu.luck.data.Token;
import me.meiallu.luck.data.TokenType;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Lexer {

    public static List<Token> lex(String path) {
        try {
            List<Token> tokenList = new ArrayList<>();
            BufferedReader bufferedReader = new BufferedReader(new FileReader(path));

            String line = bufferedReader.readLine();

            while (line != null) {
                StringBuilder builder = null;
                char[] chars = line.toCharArray();

                boolean readingString = false;

                for (Character character : chars) {
                    if (character == '"') {
                        if (!readingString) {
                            builder = new StringBuilder();
                            readingString = true;
                        } else {
                            tokenList.add(new Token(TokenType.STRING, builder.toString()));
                            builder = null;
                            readingString = false;
                        }
                        continue;
                    } else if (readingString) {
                        builder.append(character);
                        continue;
                    }

                    switch (character) {
                        case '+', '-', '*', '/', '%':
                            tokenList.add(new Token(TokenType.ARITHMETIC_OPERATOR, character));
                            break;
                        case '=':
                            tokenList.add(new Token(TokenType.ASSIGNMENT_OPERATOR, null));
                            break;
                        case ' ', ',', '(', ')':
                            if (builder != null) {
                                Token token = getToken(builder);
                                tokenList.add(token);

                                if (character == '(')
                                    token.type = TokenType.FUNCTION;

                                builder = null;
                            }

                            switch (character) {
                                case ')':
                                    tokenList.add(new Token(TokenType.FUNCTION_END, null));
                                    break;
                                case ',':
                                    tokenList.add(new Token(TokenType.SEPARATOR, null));
                                    break;
                            }
                            break;
                        default:
                            if (builder == null)
                                builder = new StringBuilder();

                            builder.append(character);
                            break;
                    }
                }

                if (builder != null)
                    tokenList.add(getToken(builder));

                if (tokenList.get(tokenList.size() - 1).type != TokenType.NEW_LINE)
                    tokenList.add(new Token(TokenType.NEW_LINE, null));

                line = bufferedReader.readLine();
            }

            bufferedReader.close();
            return tokenList;
        } catch (IOException exception) {
            throw new RuntimeException(exception);
        }
    }

    public static Token getToken(StringBuilder builder) {
        String toString = builder.toString();

        if (toString.equals("var")) {
            return new Token(TokenType.TYPE, null);
        } else {
            try {
                double parsedDouble = Double.parseDouble(toString);
                return new Token(TokenType.NUMBER, parsedDouble);
            } catch (NumberFormatException ignored) {
                return new Token(TokenType.IDENTIFIER, toString);
            }
        }
    }
}

File: src/main/kotlin/Lexer.kt
package me.meiallu.luck

import java.io.File

class Lexer {

    companion object {
        fun lex(path: String): ArrayList<Token> {
            val tokenList: ArrayList<Token> = ArrayList()

            File(path).forEachLine {
                var builder: StringBuilder? = null
                var chars = it.toCharArray()
                var readingString = false

                for (character in chars) {

                    if (character == '"') {
                        if (!readingString) {
                            builder = StringBuilder()
                            readingString = true
                        } else {
                            tokenList += Token(TokenType.STRING, builder.toString())
                            builder = null
                            readingString = false
                        }
                        continue
                    } else if (readingString) {
                        builder?.append(character)
                        continue
                    }

                    when (character) {
                        '+', '-', '*', '/', '%' -> tokenList += Token(TokenType.ARITHMETIC_OPERATOR, character)
                        '=' -> tokenList += Token(TokenType.ASSIGNMENT_OPERATOR, null)
                        ' ', ',', '(', ')' -> {
                            if (builder != null) {
                                var token = getToken(builder)
                                tokenList += token

                                if (character == '(')
                                    token.type = TokenType.FUNCTION

                                builder = null;
                            }

                            when (character) {
                                ')' -> tokenList += Token(TokenType.FUNCTION_END, null)
                                ',' -> tokenList += Token(TokenType.SEPARATOR, null)
                            }
                        }

                        else -> {
                            if (builder == null)
                                builder = StringBuilder()

                            builder.append(character)
                        }
                    }
                }

                if (builder != null)
                    tokenList += getToken(builder)

                if (tokenList[tokenList.size - 1].type != TokenType.NEW_LINE)
                    tokenList += Token(TokenType.NEW_LINE, null)
            }
            return tokenList;
        }

        private fun getToken(builder: java.lang.StringBuilder): Token {
            val toString = builder.toString()

            if (toString == "var") {
                return Token(TokenType.TYPE, null)
            } else {
                return if (toString.toDoubleOrNull() != null)
                    Token(TokenType.NUMBER, toString.toDouble())
                else
                    Token(TokenType.IDENTIFIER, toString)
            }
        }
    }
}